File: ex04/04-344.cpp
/*
 * leetcode 344
 * 
 * double-pointer: left and right
 *
 * just check the for-loop
 * the common way is just one invariable like the follows:
 * for (int i = 0; i < s.size(); i++) { ... }
 *
 */

class Solution {
public:
    void reverseString(vector<char>& s) {
        for (int i = 0, j = s.size()-1; i < s.size()/2; i++, j--) {
            swap(s[i], s[j]);
        }
    }
};

class Solution {
public:
    void reverseString(vector<char>& s) {
        int n = s.size();
        for(int i = 0, j = n-1; i < n/2; i++, j--) {
            swap(s[i], s[j]);
        }
    }
};

// recursion
class Solution {
public:
    void myReverse(int i, vector<char>& s) {
        if (i >= s.size()-i-1) return;
        swap(s[i], s[s.size()-i-1]);
        myReverse(i+1, s);
    }
    void reverseString(vector<char>& s) {
        myReverse(0, s);
    }
};

// recursion
class Solution {
public:
    void myReverse(vector<char>& s, int low, int high) {
        if (low >= high) return;
        myReverse(s, low+1, high-1);
        swap(s[low], s[high]);
    }
    void reverseString(vector<char>& s) {
        int n = s.size()-1;
        myReverse(s, 0, n);
    }
};

File: ex04/04-344.py
"""
leetcode 344
reverse string

double-pointer

pay attention to some pythonic features

"""

#
# double pointer: left and right
#
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        left, right = 0, len(s) - 1
        while left < right:
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1
        
#
# use stack
#
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        stack = []
        for char in s:
            stack.append(char)
        for i in range(len(s)):
            s[i] = stack.pop()

#
# use range
#
class Solution:
    def reverseString(self, s: List[str]) -> None:
        n = len(s)
        for i in range(n//2): # // instead of / in python
            s[i], s[n-i-1] = s[n-i-1], s[i]

